qiskit/pulse/transforms/frames.py

```diff
@@ -0,0 +1,135 @@
+
+    # Used to keep track of the frequency and phase of the channels
+    channel_trackers = {}
+    for ch in schedule.channels:
+        if isinstance(ch, chans.PulseChannel):
+            channel_trackers[ch] = ChannelTracker(ch, sample_duration)
+
+    # Add the channels that the frames broadcast on.
+    for frame in resolved_frames.values():
+        for ch in frame.channels:
+            if ch not in channel_trackers:
+                channel_trackers[ch] = ChannelTracker(ch, sample_duration)
+
+    sched = Schedule(name=schedule.name, metadata=schedule.metadata)
+
+    for time, inst in schedule.instructions:
+        if isinstance(inst, instructions.Play):
+            chan = inst.channel
+
+            if isinstance(inst.operands[0], Signal):
+                frame_idx = inst.operands[0].frame.index
+
+                if frame_idx not in resolved_frames:
+                    raise PulseError(f'{Frame(frame.index)} is not configured and cannot '
+                                     f'be resolved.')
+
+                frame = resolved_frames[frame_idx]
+
+                frame_freq = frame.frequency(time)
+                frame_phase = frame.phase(time)
+
+                # If the frequency and phase of the channel has already been set once in
+                # The past we compute shifts.
+                if channel_trackers[chan].is_initialized():
+                    freq_diff = frame_freq - channel_trackers[chan].frequency(time)
+                    phase_diff = frame_phase - channel_trackers[chan].phase(time)
+
+                    if freq_diff != 0.0:
+                        shift_freq = instructions.ShiftFrequency(freq_diff, chan)
+                        sched.insert(time, shift_freq, inplace=True)
+
+                    if phase_diff != 0.0:
+                        sched.insert(time, instructions.ShiftPhase(phase_diff, chan), inplace=True)
+
+                # If the channel's phase and frequency has not been set in the past
+                # we set it now
+                else:
+                    sched.insert(time, instructions.SetFrequency(frame_freq, chan), inplace=True)
+                    sched.insert(time, instructions.SetPhase(frame_phase, chan), inplace=True)
+
+                # Update the frequency and phase of this channel.
+                channel_trackers[chan].set_frequency(time, frame_freq)
+                channel_trackers[chan].set_phase(time, frame_phase)
+
+                play = instructions.Play(inst.operands[0].pulse, chan)
+                sched.insert(time, play, inplace=True)
+            else:
+                sched.insert(time, instructions.Play(inst.pulse, chan), inplace=True)
+
+        # Insert phase and frequency commands that are not applied to frames.
+        elif isinstance(inst, (instructions.SetFrequency, instructions.ShiftFrequency)):
+            chan = inst.channel
+
+            if issubclass(type(chan), chans.PulseChannel):
```

Comment:
This will also skip AcquireChannels, I believe. Is that intended?

What happens to set/shift phase/freq instructions on frames? Why doesn't the tracker need to be updated?

Reply:
Shift/set phase/freq instructions are taken into account by the line
```python
resolved_frame.set_frame_instructions(schedule)
```
in `def resolve_frames(schedule: Schedule, frames_config: Dict[Frame, Dict])`. This will carve out all pure frame instructions and include their effect in the `ResolvedFrame` instances.

Reply:
Skipping AcquireChannels is fine since Set/Shift instructions can only be applied to PulseChannels or Frames.